cidc_api/models/models.py

```diff
@@ -330,6 +335,58 @@ class Permissions(CommonColumns):
     trial_id = Column(String, nullable=False, index=True)
     upload_type = Column(String, nullable=False)
 
+    @with_default_session
+    def insert(self, session: Session, commit: bool = True, compute_etag: bool = True):
+        """
+        Insert this permission record into the database and add a corresponding IAM policy binding
+        on the GCS data bucket.
+
+        NOTE: values provided to the `commit` argument will be ignored. This method always commits.
+        """
+        grantee = Users.find_by_id(self.granted_to_user)
+        if grantee is None:
+            raise IntegrityError(
+                params=None,
+                statement=None,
+                orig=f"`granted_to_user` user must exist, but no user found with id {self.granted_to_user}",
+            )
+
+        # Grant IAM permission in GCS
+        grant_download_access(grantee.email, self.trial_id, self.upload_type)
+
+        try:
+            # Always commit, because we need to know whether to rollback the GCS permission grant
+            super().insert(session=session, commit=True, compute_etag=compute_etag)
+        except:
+            session.rollback()
+            # If adding the db record fails, revoke the IAM permission
+            revoke_download_access(grantee.email, self.trial_id, self.upload_type)
+            raise
+
+    @with_default_session
+    def delete(self, session: Session, commit: bool = True):
+        """
+        Delete this permission record from the database and revoke the corresponding IAM policy binding
+        on the GCS data bucket.
+
+        NOTE: values provided to the `commit` argument will be ignored. This method always commits.
+        """
+        grantee = Users.find_by_id(self.granted_to_user)
+        if grantee is None:
+            raise NoResultFound(f"no user with id {self.granted_to_user}")
+
+        # Revoke IAM permission in GCS
+        revoke_download_access(grantee.email, self.trial_id, self.upload_type)
```

Comment:
here it's in the right order. i.e. for failure scenario we at least won't have untracked IAM permissions.